esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,432 @@
+            if (sk!=StKindNone)
+                scopeTypeOverrideStatistic[sst].push_back(sk);
+            else
+            {
+                WuAttr wa = queryWuAttribute(attribs[idx2], WAMax);
+                if (wa!=WAMax)
+                    scopeTypeOverrideAttribute[sst].push_back(wa);
+                else
+                    throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name (%s) in ScopeOverrides",attribs[idx2]);
+            }
+        }
+    }
+
+    // If scope override for stats or attributes specified,
+    // then noteStatistic & noteAttribute will need to work out what to return
+    if (scopeTypeOverrideSpecified)
+    {
+
+        StringArray & attribsToReturnList = attribsToReturn.getAttributes();
+        ForEachItemIn(idx,attribsToReturnList)
+        {
+            const char * attributeName = attribsToReturnList[idx];
+            if (*attributeName==0) continue;
+
+            StatisticKind sk = queryStatisticKind(attributeName, StKindNone);
+            if (sk==StKindAll)
+                throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name in AttributeToReturn(%s)", attributeName);
+            if (sk!=StKindNone)
+            {
+                if (!returnStatisticListSpecified)
+                {
+                    returnStatisticList.set(createBitSet(StatisticFilterMaskSize,true));
+                    returnStatisticListSpecified=true;
+                }
+                returnStatisticList->set(sk, true);
+            }
+            else
+            {
+              WuAttr wa = queryWuAttribute(attributeName, WAMax);
+              if (wa!=WAMax)
+              {
+                  if (returnAttributeSpecified)
+                  {
+                      returnAttributeList.set(createBitSet(AttributeFilterMaskSize,true));
+                      returnAttributeSpecified=true;
+                  }
+                  returnAttributeList->set(wa-WANone,true);
+              }
+              else
+                  throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name in AttributeToReturn(%s)", attributeName);
+            }
+        }
+    }
+}
+
+void WUDetailsVisitor::resetScope()
+{
+    currentStatisticScopeType = SSTnone;
+    currentAttribs.clear();
+}
+
+static const char * getScopeIdFromScopeName(const char * scopeName)
+{
+    const char * p = strrchr(scopeName, ':');
+    return p?p+1:scopeName;
+}
+
+WUDetails::WUDetails(IConstWorkUnit *_workunit, const char *_wuid, IEspWUDetailsRequest &req)
+: workunit(_workunit), wuid(_wuid), requestScopeFilter(req.getFilter()),
+nestedfilter(req.getNested()), attribsToReturn(req.getAttributeToReturn()),
+scopeOptions(req.getScopeOptions()), attribOptions(req.getAttributeOptions()) { }
+
+void WUDetails::processRequest(IEspWUDetailsResponse &resp)
+{
+    buildWuScopeFilter();
+
+    WuPropertyTypes pt = PTall;
+    if (attribsToReturn.getOnlyDynamic() || *attribsToReturn.getMeasure())
```

Comment:
these lines aren't needed - can always pass PTall, but you should be calling addOutputProperties on the filter condition - otherwise the iterator cannot be optimized.

Reply:
(and if not then I need to fix my code).

Reply:
Removed.